File: src/ebus_nif.erl
-module(ebus_nif).

-define(APPNAME, ebus).
-define(LIBNAME, 'ebus').

-on_load(init/0).

-export([connection_get/2, connection_close/1, connection_unique_name/1,
         connection_request_name/3, connection_release_name/2,
         connection_add_match/2, connection_send/2, connection_call/4, connection_dispatch/1,
         connection_set_filters/2,
         connection_register_object_path/3, connection_unregister_object_path/2]).
-export([message_new_signal/3, message_new_call/4, message_new_reply/1,
         message_append_args/3, message_get_args/1, message_get_type/1,
         message_get_serial/1, message_set_serial/2, message_get_reply_serial/1,
         message_get_destination/1, message_get_path/1, message_get_interface/1, message_get_member/1,
         message_get_error/1]).
-export([watch_handle/2]).
-export([timeout_handle/1]).

%%
%% Messages
%%

-spec message_new_signal(Path::string(), IFace::string(), Name::string())
                        -> {ok, ebus:message()} | {error, term()}.
message_new_signal(_,_,_) ->
    not_loaded(?LINE).

-spec message_new_call(Dest::string(), Path::string(), IFace::string(), Name::string())
                      -> {ok, ebus:message()} | {error, term()}.
message_new_call(_,_,_,_) ->
    not_loaded(?LINE).

-spec message_new_reply(ebus:message()) -> {ok, ebus:message()} | {error, term()}.
message_new_reply(_) ->
    not_loaded(?LINE).

-spec message_append_args(ebus:message(), Signature::string(), Args::[any()]) -> ok | {error, string()}.
message_append_args(_, _, _) ->
    not_loaded(?LINE).

-spec message_get_args(ebus:message()) -> {ok, [any()]} | {error, term()}.
message_get_args(_) ->
    not_loaded(?LINE).

-spec message_get_type(ebus:message()) -> ebus:message_type().
message_get_type(_) ->
    not_loaded(?LINE).

-spec message_get_serial(ebus:message()) -> non_neg_integer().
message_get_serial(_) ->
    not_loaded(?LINE).

-spec message_set_serial(ebus:message(), non_neg_integer()) ->ok.
message_set_serial(_, _) ->
    not_loaded(?LINE).

-spec message_get_reply_serial(ebus:message()) -> non_neg_integer().
message_get_reply_serial(_) ->
    not_loaded(?LINE).

-spec message_get_destination(ebus:message()) -> string() | undefined.
message_get_destination(_) ->
    not_loaded(?LINE).

-spec message_get_path(ebus:message()) -> string() | undefined.
message_get_path(_) ->
    not_loaded(?LINE).

-spec message_get_interface(ebus:message()) -> string() | undefined.
message_get_interface(_) ->
    not_loaded(?LINE).

-spec message_get_member(ebus:message()) -> string() | undefined.
message_get_member(_) ->
    not_loaded(?LINE).

-spec message_get_error(ebus:message()) -> ok | {error, term()}.
message_get_error(_) ->
    not_loaded(?LINE).

%%
%% Connection
%%

-spec connection_get(pos_integer(), pid()) -> {ok, ebus:connection()} | {error, term()}.
connection_get(_,_) ->
    not_loaded(?LINE).

-spec connection_close(ebus:connection()) -> ok.
connection_close(_) ->
    not_loaded(?LINE).

-spec connection_unique_name(ebus:connection()) -> string().
connection_unique_name(_) ->
    not_loaded(?LINE).

-spec connection_add_match(ebus:connection(), Rule::string()) -> ok | {error, term()}.
connection_add_match(_,_) ->
    not_loaded(?LINE).

-spec connection_request_name(ebus:connection(), Name::string(), Flags::pos_integer())
                             -> {error, term()} | {ok, pos_integer()}.
connection_request_name(_, _, _) ->
    not_loaded(?LINE).

-spec connection_release_name(ebus:connection(), Name::string()) -> {ok, pos_integer()} | {error, term()}.
connection_release_name(_, _) ->
    not_loaded(?LINE).

-spec connection_send(ebus:connection(), ebus:message()) -> ok | {error, enomem}.
connection_send(_,_) ->
    not_loaded(?LINE).

-spec connection_call(ebus:connection(), ebus:message(), pid(), integer()) -> ok | {error, enomem}.
connection_call(_,_,_,_) ->
    not_loaded(?LINE).

-spec connection_dispatch(ebus:connection()) -> non_neg_integer().
connection_dispatch(_) ->
    not_loaded(?LINE).

-spec connection_set_filters(ebus:connection(), [{Ref::reference(), ebus:filter()}]) -> ok | {error, enomem}.
connection_set_filters(_,_) ->
    not_loaded(?LINE).

-spec connection_register_object_path(ebus:connection(), string(), pid()) -> ok | {error, enomem | already}.
connection_register_object_path(_,_,_) ->
    not_loaded(?LINE).

-spec connection_unregister_object_path(ebus:connection(), string()) -> ok | {error, enomem | already}.
connection_unregister_object_path(_,_) ->
    not_loaded(?LINE).


%%
%% Watches
%%

-spec watch_handle(ebus:watch(), non_neg_integer()) -> boolean().
watch_handle(_,_) ->
    not_loaded(?LINE).

%%
%% Timeout
%%

-spec timeout_handle(reference()) -> ok.
timeout_handle(_) ->
    not_loaded(?LINE).

init() ->
    SoName = case code:priv_dir(?APPNAME) of
        {error, bad_name} ->
            case filelib:is_dir(filename:join(["..", priv])) of
                true ->
                    filename:join(["..", priv, ?LIBNAME]);
                _ ->
                    filename:join([priv, ?LIBNAME])
            end;
        Dir ->
            filename:join(Dir, ?LIBNAME)
    end,
    erlang:load_nif(SoName, 0).

not_loaded(Line) ->
    erlang:nif_error({not_loaded, [{module, ?MODULE}, {line, Line}]}).

File: c_src/ebus_nif.c
/*
 * Copyright 2018 Helium Systems Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "ebus_connection.h"
#include "ebus_message.h"
#include "ebus_shared.h"
#include "ebus_error.h"
#include "ebus_timeout.h"
#include "ebus_watch.h"
#include "ebus_object.h"

ERL_NIF_TERM ATOM_OK;
ERL_NIF_TERM ATOM_ERROR;
ERL_NIF_TERM ATOM_ENOMEM;
ERL_NIF_TERM ATOM_TRUE;
ERL_NIF_TERM ATOM_FALSE;
ERL_NIF_TERM ATOM_UNDEFINED;
ERL_NIF_TERM ATOM_ALREADY;

static ErlNifFunc nif_funcs[] = {
    {"connection_get", 2, ebus_connection_get, ERL_NIF_DIRTY_JOB_IO_BOUND},
    {"connection_close", 1, ebus_connection_close, ERL_NIF_DIRTY_JOB_IO_BOUND},
    {"connection_unique_name", 1, ebus_connection_unique_name, 0},
    {"connection_request_name", 3, ebus_connection_request_name, 0},
    {"connection_release_name", 2, ebus_connection_release_name, 0},
    {"connection_add_match", 2, ebus_connection_add_match, ERL_NIF_DIRTY_JOB_IO_BOUND},
    {"connection_send", 2, ebus_connection_send, ERL_NIF_DIRTY_JOB_IO_BOUND},
    {"connection_call", 4, ebus_connection_call, ERL_NIF_DIRTY_JOB_IO_BOUND},
    {"connection_dispatch", 1, ebus_connection_dispatch, ERL_NIF_DIRTY_JOB_CPU_BOUND},
    {"connection_set_filters", 2, ebus_connection_set_filters, ERL_NIF_DIRTY_JOB_CPU_BOUND},
    {"connection_register_object_path", 3, ebus_connection_register_object_path, 0},
    {"connection_unregister_object_path", 2, ebus_connection_unregister_object_path, 0},

    {"message_new_signal", 3, ebus_message_new_signal, 0},
    {"message_new_call", 4, ebus_message_new_call, 0},
    {"message_new_reply", 1, ebus_message_new_reply, 0},
    {"message_append_args", 3, ebus_message_append_args, ERL_NIF_DIRTY_JOB_CPU_BOUND},
    {"message_get_args", 1, ebus_message_get_args, ERL_NIF_DIRTY_JOB_CPU_BOUND},
    {"message_get_type", 1, ebus_message_get_type, 0},
    {"message_get_serial", 1, ebus_message_get_serial, 0},
    {"message_set_serial", 2, ebus_message_set_serial, 0},
    {"message_get_reply_serial", 1, ebus_message_get_reply_serial, 0},
    {"message_get_destination", 1, ebus_message_get_destination, 0},
    {"message_get_path", 1, ebus_message_get_path, 0},
    {"message_get_interface", 1, ebus_message_get_interface, 0},
    {"message_get_member", 1, ebus_message_get_member, 0},
    {"message_get_error", 1, ebus_message_get_error, 0},
    {"message_infer_signature", 1, ebus_message_infer_signature, 0},

    {"watch_handle", 2, ebus_watch_handle, ERL_NIF_DIRTY_JOB_IO_BOUND},

    {"timeout_handle", 1, ebus_timeout_handle, ERL_NIF_DIRTY_JOB_CPU_BOUND},
};

static int
load(ErlNifEnv * env, void ** priv_data, ERL_NIF_TERM load_info)
{
    (void)priv_data;
    (void)load_info;


    ATOM(ATOM_OK, "ok");
    ATOM(ATOM_ERROR, "error");
    ATOM(ATOM_ENOMEM, "enomem");
    ATOM(ATOM_TRUE, "true");
    ATOM(ATOM_FALSE, "false");
    ATOM(ATOM_UNDEFINED, "undefined");
    ATOM(ATOM_ALREADY, "already");

    ebus_connection_load(env);
    ebus_message_load(env);
    ebus_watch_load(env);
    ebus_timeout_load(env);
    ebus_filter_load(env);
    ebus_object_load(env);
    ebus_error_load(env);

    return 0;
}

ERL_NIF_INIT(ebus_nif, nif_funcs, load, NULL, NULL, NULL);
